lib/content-negotiation.js

```diff
@@ -0,0 +1,93 @@
+// Copyright IBM Corp. 2016. All Rights Reserved.
+// Node module: strong-error-handler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+var accepts = require('accepts');
+var debug = require('debug')('strong-error-handler:http-response');
+var parseParam = require('qs').parse;
+var parseUrl = require('parseurl');
+var sendJson = require('./send-json');
+var sendHTML = require('./send-html');
+
+module.exports = contentNegotiation;
+
+/**
+ * Handles req.accepts and req.query._format and options.defaultTypes
+ * to resolve the correct operation
+ *
+ * @param req request object
+ * @param {Object} options options of strong-error-handler
+ * @returns {Function} Opeartion function with signature `fn(res, data)`
+ */
+function contentNegotiation(req, options) {
+  options = options || {};
+
+  var DEFAULT_SUPPORTED_TYPES = [
+    'application/json', 'application/javascript', 'application/xml',
+    'text/javascript', 'text/html', 'text/xml',
+    'json', 'xml', 'html',
+  ];
+  var defaultType = options.defaultType || 'json';
+
+  // checking if user provided defaultType is supported
+  if (options.defaultType &&
+    DEFAULT_SUPPORTED_TYPES.indexOf(options.defaultType) > -1) {
+    debug('Accepting options.defaultType `%s`', options.defaultType);
+    defaultType = options.defaultType;
+  }
+
+  // decide to use resolvedType or defaultType
+  // Please note that accepts assumes the order of content-type is provided
+  // in the priority returned
+  // example
+  // Content-type: text/html, */*, application/json ---> will resolve as text/html
+  // Content-type: application/json, */*, text/html ---> will resolve as application/json
+  // Content-type: */*, application/json, text/html ---> will resolve as */*
+  // Chrome accepts defaults to `text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*`
+  // which in our case will prioritize text/html because its first
+  var resolvedContentType = accepts(req).types(DEFAULT_SUPPORTED_TYPES);
+  debug('Resolved content-type', resolvedContentType);
+  var contentType = resolvedContentType || defaultType;
+
+  // to receive _format from user's url param to overide the content type
+  // req.query (eg /api/Users/1?_format=json will overide content negotiation
+  // https://github.com/strongloop/strong-remoting/blob/ac3093dcfbb787977ca0229b0f672703859e52e1/lib/http-context.js#L643-L645
+  var queryString = parseUrl(req).query;
+  var query = parseParam(queryString);
```

Comment:
In most application, the query string will be already parsed for you and accessible via `req.query`, see http://expressjs.com/en/4x/api.html#req.query.

For applications that don't setup query-parsing middleware, I think it's ok to ignore `_format` in the query string.

In other words, I am proposing to simplify these two lines into:

```js
var query = req.query || {};
```